File: backend/matching/scoring.py
from __future__ import annotations

from typing import Any

from .embeddings import semantic_similarity


WEIGHTS = {
    "semantic": 0.40,
    "course": 0.30,
    "interest": 0.20,
    "free_time": 0.10,
}


def rank_matches(query_student: Any, candidates: list[Any], top_k: int = 20) -> list[dict]:
    scored = []
    for candidate in candidates:
        features = compute_features(query_student, candidate)
        score = round(
            100
            * (
                WEIGHTS["semantic"] * features["semantic_score"]
                + WEIGHTS["course"] * features["course_score"]
                + WEIGHTS["interest"] * features["interest_score"]
                + WEIGHTS["free_time"] * features["free_time_score"]
            )
        )
        reasons = explain(query_student, candidate)
        if score == 0 or not reasons:
            continue
        scored.append(
            {
                "studentId": candidate.id,
                "score": min(100, score),
                "reasons": reasons,
                "aiSummary": summarize(query_student.name, candidate.name, reasons),
                "debug": features,
            }
        )

    scored.sort(key=lambda item: item["score"], reverse=True)
    return [{key: value for key, value in item.items() if key != "debug"} for item in scored[:top_k]]


def compute_features(query_student: Any, candidate: Any) -> dict[str, float]:
    shared_courses = _shared_courses(query_student, candidate)
    shared_interests = _shared_interests(query_student, candidate)
    shared_free_slots = _shared_free_slots(query_student, candidate)

    return {
        "semantic_score": semantic_similarity(query_student, candidate),
        "course_score": min(1.0, len(shared_courses) / 2),
        "interest_score": min(1.0, len(shared_interests) / 3),
        "free_time_score": 1.0 if shared_free_slots else 0.0,
    }


def explain(query_student: Any, candidate: Any) -> list[dict]:
    reasons = []

    for course in _shared_courses(query_student, candidate)[:3]:
        reasons.append(
            {
                "type": "shared_course",
                "courseCode": course["code"],
                "courseName": course["name"],
            }
        )

    for interest in _shared_interests(query_student, candidate)[:3]:
        reasons.append({"type": "shared_interest", "interest": interest})

    shared_free_slots = _shared_free_slots(query_student, candidate)
    if shared_free_slots:
        reasons.append(
            {
                "type": "shared_free_time",
                "day": shared_free_slots[0]["day"],
                "window": shared_free_slots[0]["window"],
            }
        )

    return reasons


def summarize(student_a_name: str, student_b_name: str, reasons: list[dict]) -> str:
    courses = [reason["courseCode"] for reason in reasons if reason["type"] == "shared_course"]
    interests = [reason["interest"] for reason in reasons if reason["type"] == "shared_interest"]
    free_time = next((reason for reason in reasons if reason["type"] == "shared_free_time"), None)
    parts = []

    if courses:
        parts.append(f"both take {_join_human(courses)}")
    if interests:
        parts.append(f"share {_join_human(interests)}")
    if free_time:
        parts.append(f"are free {free_time['day']} {free_time['window']}")

    if not parts:
        return f"{student_a_name} and {student_b_name} have overlapping campus signals."
    return f"{student_a_name} and {student_b_name} {_join_human(parts)}."


def _shared_courses(query_student: Any, candidate: Any) -> list[dict[str, str]]:
    candidate_courses = {course.code.upper(): course for course in candidate.courses}
    shared = []
    for course in query_student.courses:
        if course.code.upper() in candidate_courses:
            shared.append({"code": course.code, "name": course.name})
    return shared


def _shared_interests(query_student: Any, candidate: Any) -> list[str]:
    candidate_interests = set(candidate.interests)
    return [interest for interest in query_student.interests if interest in candidate_interests]


def _shared_free_slots(query_student: Any, candidate: Any) -> list[dict[str, str]]:
    candidate_slots = {(slot.day, slot.window) for slot in candidate.freeSlots}
    return [
        {"day": slot.day, "window": slot.window}
        for slot in query_student.freeSlots
        if (slot.day, slot.window) in candidate_slots
    ]


def _join_human(items: list[str]) -> str:
    if len(items) == 0:
        return ""
    if len(items) == 1:
        return items[0]
    if len(items) == 2:
        return f"{items[0]} and {items[1]}"
    return f"{', '.join(items[:-1])}, and {items[-1]}"

File: src/lib/matching/scoring.ts
// [Owner: A] Explainable scoring algorithm — this is the feature that
// differentiates the app from a generic dating-app clone. Never return a
// bare number: every point of `score` must trace back to a MatchReason.
import type { MatchReason } from "@/lib/types";

const CATEGORY_WEIGHTS = {
  courses: 45,
  interests: 35,
  freeTime: 20,
};

export interface ScoringInput {
  studentACourses: { code: string; name: string }[];
  studentBCourses: { code: string; name: string }[];
  studentAInterests: string[];
  studentBInterests: string[];
  studentAFreeSlots: { day: string; window: string }[];
  studentBFreeSlots: { day: string; window: string }[];
}

export interface ScoringOutput {
  score: number;
  reasons: MatchReason[];
}

export function computeMatchScore(input: ScoringInput): ScoringOutput {
  const reasons: MatchReason[] = [];

  // Shared courses are matched by course code only. Course names are display
  // metadata because imports may disagree on wording.
  const bCourseCodes = new Set(input.studentBCourses.map((course) => course.code.toUpperCase()));
  const sharedCourses: MatchReason[] = [];
  for (const course of input.studentACourses) {
    if (bCourseCodes.has(course.code.toUpperCase())) {
      sharedCourses.push({ type: "shared_course", courseCode: course.code, courseName: course.name });
    }
  }

  // Shared interests
  const bInterests = new Set(input.studentBInterests);
  const sharedInterests: MatchReason[] = [];
  for (const interest of input.studentAInterests) {
    if (bInterests.has(interest)) {
      sharedInterests.push({ type: "shared_interest", interest });
    }
  }

  // Shared free time is capped as one category so empty timetable windows do
  // not swamp course/profile mismatch and force every match to 100%.
  const bSlots = new Set(input.studentBFreeSlots.map((s) => `${s.day}|${s.window}`));
  const sharedFreeSlots: MatchReason[] = [];
  for (const slot of input.studentAFreeSlots) {
    if (bSlots.has(`${slot.day}|${slot.window}`)) {
      sharedFreeSlots.push({ type: "shared_free_time", day: slot.day, window: slot.window });
    }
  }

  reasons.push(...sharedCourses, ...sharedInterests, ...sharedFreeSlots.slice(0, 3));
  reasons.sort((a, b) => weightOf(b) - weightOf(a));

  const courseDenominator = Math.max(input.studentACourses.length, input.studentBCourses.length, 1);
  const interestDenominator = Math.max(input.studentAInterests.length, input.studentBInterests.length, 1);
  const courseScore = CATEGORY_WEIGHTS.courses * (sharedCourses.length / courseDenominator);
  const interestScore = CATEGORY_WEIGHTS.interests * (sharedInterests.length / interestDenominator);
  const freeTimeScore = CATEGORY_WEIGHTS.freeTime * Math.min(1, sharedFreeSlots.length / 3);

  return { score: Math.min(100, Math.round(courseScore + interestScore + freeTimeScore)), reasons };
}

function weightOf(reason: MatchReason): number {
  if (reason.type === "shared_course") return CATEGORY_WEIGHTS.courses;
  if (reason.type === "shared_interest") return CATEGORY_WEIGHTS.interests;
  return CATEGORY_WEIGHTS.freeTime;
}
